brambling/tests/functional/test_webhooks.py

```diff
@@ -144,3 +159,229 @@ def test_confirmation(self):
+
+        with self.assertRaises(Http404):
+            self.view(self.request)
+
+
+class SuccessfulRefundWebhookTestCase(TestCase):
+    @vcr.use_cassette(os.path.join(VCR_DIR, 'stripe_webhook_refund_success.yaml'))
+    def setUp(self):
+        stripe_prep(TEST)
+
+        self.event = EventFactory()
+        self.order = OrderFactory(event=self.event)
+
+        self.factory = RequestFactory()
+        self.view = StripeWebhookView.as_view()
+        self.stripe_event = {'id': 'evt_123_event_id'}
+        self.data = json.dumps(self.stripe_event)
+
+        self.request = self.factory.post(path='/',
+                                         content_type='application/json',
+                                         data=self.data)
+
+        item = ItemFactory(event=self.event)
+        item_option = ItemOptionFactory(price=60, item=item)
+        self.order.add_to_cart(item_option)
+
+        token = stripe.Token.create(
+            card={
+                'number': '4242424242424242',
+                'exp_month': 12,
+                'exp_year': 2017,
+                'cvc': '123'
+            },
+        )
+
+        charge = stripe_charge(token, 100, self.order, self.event)
+
+        self.txn = Transaction.from_stripe_charge(
+            charge,
+            event=self.event,
+            order=self.order,
+        )
+
+        self.refund = stripe_refund(self.order, self.event, charge.id, 100)
+
+        data = mock.Mock(object=mock.Mock(name='charge', id=charge.id))
+        self.mock_event = mock.Mock(
+            data=data, type='charge.refunded', livemode=False)
+        self.order.mark_cart_paid(self.txn)
+
+    @mock.patch('stripe.Event.retrieve')
+    def test_transaction_refunded_successfully(self, event_retrieve):
+        event_retrieve.return_value = self.mock_event
+        response = self.view(self.request)
+
+        refund = Transaction.objects.get(
+            related_transaction=self.txn,
+            transaction_type=Transaction.REFUND,
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(refund.amount, -100)
+        self.assertEqual(refund.application_fee, -2.5)
+        self.assertEqual(refund.method, Transaction.STRIPE)
+        self.assertEqual(refund.order, self.order)
+        self.assertEqual(refund.event, self.event)
+
+    @mock.patch('stripe.Event.retrieve')
+    def test_events_should_be_processed_exactly_once(self, event_retrieve):
+        event_retrieve.return_value = self.mock_event
+        response1 = self.view(self.request)
+        response2 = self.view(self.request)
+
+        self.assertEqual(response1.status_code, 200)
+        self.assertEqual(response2.status_code, 200)
+        Transaction.objects.get(related_transaction=self.txn,
+                                transaction_type=Transaction.REFUND)
+
+    @mock.patch('stripe.Event.retrieve')
+    def test_events_should_be_processed_exactly_once_in_livemode(self, event_retrieve):
```

Comment:
Since stripe_prep() isn't run again and self.txn isn't updated, this would seem to imply that we don't care if there's a mismatch between the event livemode and whether the Transaction is live or test.

Potential issue?

Reply:
self.txn is now updated.